servant-server/src/Servant/Server/Internal.hs

```diff
@@ -134,8 +135,8 @@ processMethodRouter :: forall a. ConvertibleStrings a B.ByteString
                     -> Maybe [(HeaderName, B.ByteString)]
                     -> Request -> RouteResult Response
 processMethodRouter handleA status method headers request = case handleA of
-  Nothing -> failWith UnsupportedMediaType
-  Just (contentT, body) -> succeedWith $ responseLBS status hdrs bdy
+  Nothing -> FailFatal err406
```

Comment:
I don't see why this should be fatal. In my opinion,
```
Put '[Foo, Bar] T
```
should be equivalent to
```
     Put '[Foo] T
:<|> Put '[Bar] T
```
for reasons of modularity. This doesn't seem to be supported if we make the failure fatal at this point.